File: Framework/DataSynchronizationManager.cs
using System.Collections.Generic;
using System.Management.Instrumentation;
using Framework.Data_Manipulation;
using Framework.Domain;

namespace Framework
{

    public interface IDataSynchronizationManager
    {
        void RegisterEntity<TEntity>(IBaseMapper<TEntity> mapper, IBaseQueryObject<TEntity> queryObject)
            where TEntity : IDomainObject;
        IRepository<TEntity> GetRepository<TEntity>()
            where TEntity : IDomainObject;
        IBaseQueryObject<TEntity> GetQueryObject<TEntity>()
            where TEntity : IDomainObject;
        IBaseMapper<TEntity> GetMapper<TEntity>()
            where TEntity : IDomainObject;
        IList<TEntity> GetLoadedEntities<TEntity>()
            where TEntity : IDomainObject;
    }

    public class DataSynchronizationManager : IDataSynchronizationManager
    {
        static IDataSynchronizationManager s_instance = new DataSynchronizationManager();

        IDictionary<string, object> _serviceContainerDictionary = new Dictionary<string, object>();

        private DataSynchronizationManager()
        {
        }

        public static IDataSynchronizationManager GetInstance()
        {
            return s_instance;
        }

        string GetServiceContainerKey<TEntity>()
            where TEntity : IDomainObject
        {
            return typeof(TEntity).Name;
        }

        public void RegisterEntity<TEntity>(IBaseMapper<TEntity> mapper, IBaseQueryObject<TEntity> queryObject)
            where TEntity : IDomainObject
        {
            IEntityServiceContainer<TEntity> serviceContainer = new EntityServiceContainer<TEntity>
            {
                LoadedEntities = new List<TEntity>(),
                Repository = new Repository<TEntity>(),
                Mapper = mapper,
                QueryObject = queryObject
            };

            string key = GetServiceContainerKey<TEntity>();

            if (ServiceContainerExists(key))
                _serviceContainerDictionary.Remove(key);

            _serviceContainerDictionary.Add(key, serviceContainer);
        }

        bool ServiceContainerExists<TEntity>()
            where TEntity : IDomainObject
        {
            return _serviceContainerDictionary.ContainsKey(GetServiceContainerKey<TEntity>());
        }

        bool ServiceContainerExists(string key)
        {
            return _serviceContainerDictionary.ContainsKey(key);
        }

        IEntityServiceContainer<TEntity> GetServiceContainer<TEntity>()
            where TEntity : IDomainObject
        {
            IEntityServiceContainer<TEntity> serviceContainer = (IEntityServiceContainer<TEntity>)_serviceContainerDictionary[GetServiceContainerKey<TEntity>()];

            if (serviceContainer == null)
            {
                string key = GetServiceContainerKey<TEntity>(); 
                
                throw new InstanceNotFoundException(string.Format("Service container with key '{0}' not found.", key));
            }

            return serviceContainer;
        }

        public IRepository<TEntity> GetRepository<TEntity>()
            where TEntity : IDomainObject
        {
            return GetServiceContainer<TEntity>().Repository;
        }

        public IBaseQueryObject<TEntity> GetQueryObject<TEntity>()
            where TEntity : IDomainObject
        {
            return GetServiceContainer<TEntity>().QueryObject;
        }

        public IBaseMapper<TEntity> GetMapper<TEntity>()
           where TEntity : IDomainObject
        {
            return GetServiceContainer<TEntity>().Mapper;
        }

        public IList<TEntity> GetLoadedEntities<TEntity>()
           where TEntity : IDomainObject
        {
            return GetServiceContainer<TEntity>().LoadedEntities;
        }
    }
}

File: Android/framework/src/main/java/com/DataSynchronizationManager.java
package com;

import com.Interfaces.IDataSynchronizationManager;
import com.Interfaces.IRepository;
import com.datamanipulation.BaseMapperInterfaces.IBaseMapperConcrete;
import com.datamanipulation.BaseQueryObjectInterfaces.IBaseQueryObjectConcrete;
import com.domain.DomainObjectInterfaces.IDomainObject;

import java.lang.reflect.Field;
import java.rmi.NoSuchObjectException;
import java.util.HashMap;
import java.util.List;

/**
 * Created by aiko on 3/4/17.
 */

public class DataSynchronizationManager implements IDataSynchronizationManager{

    static IDataSynchronizationManager s_instance = new DataSynchronizationManager();

    HashMap<String, Object> _serviceContainerDictionary = new HashMap<String, Object>();

    public static IDataSynchronizationManager GetInstance()
    {
        return s_instance;
    }

    public <TEntity extends IDomainObject> void RegisterEntity(Class<TEntity> thisClass, IBaseMapperConcrete<TEntity> mapper, List<IBaseQueryObjectConcrete<TEntity>> queryList)
    {
        EntityServiceContainer<TEntity> serviceContainer = new EntityServiceContainer<TEntity>();

        serviceContainer.Mapper = mapper;
        serviceContainer.IdentityMap = new IdentityMap(thisClass);
        serviceContainer.Repository = new Repository(thisClass, this);
        serviceContainer.QueryDictionary = ConvertQueryListToDictionary(queryList);
        serviceContainer.PrimitiveFields = GetPrimitiveFields(thisClass);


        String key = GetServiceContainerKey(thisClass);

        if (ServiceContainerExists(key))
            _serviceContainerDictionary.remove(key);

        _serviceContainerDictionary.put(key, serviceContainer);
    }

    <TEntity extends IDomainObject>  HashMap<String, IBaseQueryObjectConcrete<TEntity>> ConvertQueryListToDictionary(List<IBaseQueryObjectConcrete<TEntity>> queryList)
    {
        HashMap<String, IBaseQueryObjectConcrete<TEntity>>  queryDictionary = new HashMap<String, IBaseQueryObjectConcrete<TEntity>>();

        if ((queryList == null) || (queryList.size() < 1))
            return queryDictionary;

        for (int index = 0; index < queryList.size(); index++)
        {
            IBaseQueryObjectConcrete<TEntity> query = queryList.get(index);

            if(query == null)
                continue;

            queryDictionary.put(query.GetSearchInputType().getName(), query);
        }

        return queryDictionary;
    }

    <TEntity extends IDomainObject>  HashMap<String, Field> GetPrimitiveFields(Class<TEntity> thisClass)
    {
        HashMap<String, Field> fieldMap = new HashMap<String, Field>();
        Field []fields = thisClass.getDeclaredFields();

        if((fields == null) || (fields.length < 1))
            return fieldMap;

        for(int index = 0; index < fields.length; index++){
            Field field = fields[index];

            if((!field.getType().isAssignableFrom(String.class)) && (!field.getType().isPrimitive()))
                continue;

            fieldMap.put(field.getName(), field);
        }

        return fieldMap;
    }

    <TEntity extends IDomainObject> String GetServiceContainerKey(Class<TEntity> thisClass)
    {
        return thisClass.getName();
    }

    boolean ServiceContainerExists(String key)
    {
        return _serviceContainerDictionary.containsKey(key);
    }

    public <TEntity extends IDomainObject> IRepository<TEntity> GetRepository(Class<TEntity> thisClass)
    {
        EntityServiceContainer<TEntity> container = null;

        try
        {
            container = GetServiceContainer(thisClass);
        }
        catch (NoSuchObjectException ex){
        }

        return container.Repository;
    }

    <TEntity extends IDomainObject> EntityServiceContainer<TEntity> GetServiceContainer(Class<TEntity> thisClass)
            throws NoSuchObjectException
    {
        String key = GetServiceContainerKey(thisClass);

        if (!ServiceContainerExists(key))
            throw new NoSuchObjectException(String.format("Service container with key '%s' not found.", key));

        EntityServiceContainer<TEntity> serviceContainer = (EntityServiceContainer<TEntity>)_serviceContainerDictionary.get(key);

        return serviceContainer;
    }
}
